cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +242,356 @@ impl UpdateConfig for Api {
-            x509.set_subject_name(subject_name)
-                .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-            x509.set_pubkey(&x509_req_public_key)
-                .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-            x509.set_not_before(
-                &*openssl::asn1::Asn1Time::days_from_now(0)
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?,
-            )
-            .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-            // Copy extensions from x509_req to the new cert.
-            let req_extensions = x509_req.extensions();
-
-            // x509_req.extensions() returns an Err variant if no extensions are present in the req.
-            // Ignore this Err and only copy extensions if provided in the req.
-            if let Ok(req_extensions) = req_extensions {
-                for extension in req_extensions {
-                    x509.append_extension(extension)
-                        .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-                }
-            }
-
-            let issuer_private_key = std::ffi::CString::new(issuer_private_key.0.clone())
-                .map_err(|err| Error::invalid_parameter("issuer.privateKeyHandle", err))?;
-            let issuer_private_key = api
-                .key_engine
-                .load_private_key(&issuer_private_key)
-                .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+                Ok(())
+            })
+            .ok()
+            .transpose()?;
+
+        let issuer_x509_stack: Vec<X509>;
+        let (issuer_name, min_expiry, issuer_pem) = if &id != &issuer.0 {
+            let issuer_x509_pem = get_cert_inner(
+                    &api.homedir_path,
+                    &api.preloaded_certs,
+                    &issuer.0
+                )?
+                .ok_or_else(|| Error::invalid_parameter("issuer.certId", "not found"))?;
+            issuer_x509_stack = X509::stack_from_pem(&issuer_x509_pem)?;
 
-            let x509 = if issuer_id == id {
-                // Issuer is the same as the cert being created, which means the caller wants the cert to be self-signed.
+            let issuer_x509 = issuer_x509_stack.get(0)
+                .ok_or_else(|| Error::invalid_parameter("issuer.certId", "invalid issuer"))?;
 
-                x509.set_not_after(&not_after)
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+            let issuer_expiry = issuer_x509.not_after();
 
-                x509.set_issuer_name(subject_name)
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+            ( issuer_x509.subject_name()
+            , if expiry < issuer_expiry { expiry } else { issuer_expiry }
+            , Some(issuer_x509_pem)
+            )
```

Comment:
```suggestion
            (
                issuer_x509.subject_name(),
                if expiry < issuer_expiry { expiry } else { issuer_expiry },
                Some(issuer_x509_pem),
            )
```

Reply:
Fixed in e2d57edbe5898347cf7633f3a59c841737c9e963.